# The Worst Hitter in MLB History

In [1]:
import pandas as pd
batting = pd.read_csv('../data/Batting.csv')

Data courtesy of [Sean Lahman's Baseball](http://www.seanlahman.com/baseball-archive/statistics/)

In [2]:
import scipy.stats as stats
import matplotlib.pyplot as plt
import numpy as np

## Exploring the Batting Dataset

In [3]:
batting.head()

playerID  yearID  stint teamID lgID   G   AB   R   H  2B  ...   RBI   SB  \
0  abercda01    1871      1    TRO  NaN   1    4   0   0   0  ...   0.0  0.0   
1   addybo01    1871      1    RC1  NaN  25  118  30  32   6  ...  13.0  8.0   
2  allisar01    1871      1    CL1  NaN  29  137  28  40   4  ...  19.0  3.0   
3  allisdo01    1871      1    WS3  NaN  27  133  28  44  10  ...  27.0  1.0   
4  ansonca01    1871      1    RC1  NaN  25  120  29  39  11  ...  16.0  6.0   

    CS  BB   SO  IBB  HBP  SH  SF  GIDP  
0  0.0   0  0.0  NaN  NaN NaN NaN   NaN  
1  1.0   4  0.0  NaN  NaN NaN NaN   NaN  
2  1.0   2  5.0  NaN  NaN NaN NaN   NaN  
3  1.0   0  2.0  NaN  NaN NaN NaN   NaN  
4  2.0   2  1.0  NaN  NaN NaN NaN   NaN  

[5 rows x 22 columns]

In [4]:
batting.columns

Index(['playerID', 'yearID', 'stint', 'teamID', 'lgID', 'G', 'AB', 'R', 'H',
       '2B', '3B', 'HR', 'RBI', 'SB', 'CS', 'BB', 'SO', 'IBB', 'HBP', 'SH',
       'SF', 'GIDP'],
      dtype='object')

In [5]:
batting.tail()

playerID  yearID  stint teamID lgID    G   AB   R    H  2B  ...  \
102811  zimmejo02    2016      1    DET   AL   19    4   0    1   0  ...   
102812  zimmery01    2016      1    WAS   NL  115  427  60   93  18  ...   
102813  zobribe01    2016      1    CHN   NL  147  523  94  142  31  ...   
102814  zuninmi01    2016      1    SEA   AL   55  164  16   34   7  ...   
102815   zychto01    2016      1    SEA   AL   12    0   0    0   0  ...   

         RBI   SB   CS  BB     SO  IBB  HBP   SH   SF  GIDP  
102811   0.0  0.0  0.0   0    2.0  0.0  0.0  1.0  0.0   0.0  
102812  46.0  4.0  1.0  29  104.0  1.0  5.0  0.0  6.0  12.0  
102813  76.0  6.0  4.0  96   82.0  6.0  4.0  4.0  4.0  17.0  
102814  31.0  0.0  0.0  21   65.0  0.0  6.0  0.0  1.0   0.0  
102815   0.0  0.0  0.0   0    0.0  0.0  0.0  0.0  0.0   0.0  

[5 rows x 22 columns]

### Checking out Pete Rose's Career Numbers

In [6]:
batting[batting['playerID'] == 'rosepe01']

playerID  yearID  stint teamID lgID    G   AB    R    H  2B  ...   RBI  \
43287  rosepe01    1963      1    CIN   NL  157  623  101  170  25  ...  41.0   
44033  rosepe01    1964      1    CIN   NL  136  516   64  139  13  ...  34.0   
44793  rosepe01    1965      1    CIN   NL  162  670  117  209  35  ...  81.0   
45567  rosepe01    1966      1    CIN   NL  156  654   97  205  38  ...  70.0   
46350  rosepe01    1967      1    CIN   NL  148  585   86  176  32  ...  76.0   
47085  rosepe01    1968      1    CIN   NL  149  626   94  210  42  ...  49.0   
47978  rosepe01    1969      1    CIN   NL  156  627  120  218  33  ...  82.0   
48920  rosepe01    1970      1    CIN   NL  159  649  120  205  37  ...  52.0   
49797  rosepe01    1971      1    CIN   NL  160  632   86  192  27  ...  44.0   
50693  rosepe01    1972      1    CIN   NL  154  645  107  198  31  ...  57.0   
51575  rosepe01    1973      1    CIN   NL  160  680  115  230  36  ...  64.0   
52484  rosepe01    1974      1    CIN   NL  163  652  110  185  45  ...  51.0   
53387  rosepe01    1975      1    CIN   NL  162  662  112  210  47  ...  74.0   
54275  rosepe01    1976      1    CIN   NL  162  665  130  215  42  ...  63.0   
55227  rosepe01    1977      1    CIN   NL  162  655   95  204  38  ...  64.0   
56194  rosepe01    1978      1    CIN   NL  159  655  103  198  51  ...  52.0   
57157  rosepe01    1979      1    PHI   NL  163  628   90  208  40  ...  59.0   
58112  rosepe01    1980      1    PHI   NL  162  655   95  185  42  ...  64.0   
59047  rosepe01    1981      1    PHI   NL  107  431   73  140  18  ...  33.0   
60022  rosepe01    1982      1    PHI   NL  162  634   80  172  25  ...  54.0   
61029  rosepe01    1983      1    PHI   NL  151  493   52  121  14  ...  45.0   
61999  rosepe01    1984      1    MON   NL   95  278   34   72   6  ...  23.0   
62000  rosepe01    1984      2    CIN   NL   26   96    9   35   9  ...  11.0   
62992  rosepe01    1985      1    CIN   NL  119  405   60  107  12  ...  46.0   
64015  rosepe01    1986      1    CIN   NL   72  237   15   52   8  ...  25.0   

         SB    CS   BB    SO   IBB   HBP   SH   SF  GIDP  
43287  13.0  15.0   55  72.0   0.0   5.0  6.0  6.0   8.0  
44033   4.0  10.0   36  51.0   0.0   2.0  3.0  1.0   6.0  
44793   8.0   3.0   69  76.0   2.0   8.0  8.0  2.0  10.0  
45567   4.0   9.0   37  61.0   3.0   1.0  7.0  1.0  12.0  
46350  11.0   6.0   56  66.0   9.0   3.0  1.0  2.0   9.0  
47085   3.0   7.0   56  76.0  15.0   4.0  2.0  4.0  11.0  
47978   7.0  10.0   88  65.0  18.0   5.0  2.0  6.0  13.0  
48920  12.0   7.0   73  64.0  10.0   2.0  0.0  4.0   7.0  
49797  13.0   9.0   68  50.0  15.0   3.0  1.0  3.0   9.0  
50693  10.0   3.0   73  46.0   4.0   7.0  2.0  2.0   7.0  
51575  10.0   7.0   65  42.0   6.0   6.0  1.0  0.0  14.0  
52484   2.0   4.0  106  54.0  14.0   5.0  1.0  6.0   9.0  
53387   0.0   1.0   89  50.0   8.0  11.0  1.0  1.0  13.0  
54275   9.0   5.0   86  54.0   7.0   6.0  0.0  2.0  17.0  
55227  16.0   4.0   66  42.0   7.0   5.0  1.0  4.0   9.0  
56194  13.0   9.0   62  30.0   6.0   3.0  2.0  7.0   8.0  
57157  20.0  11.0   95  32.0  10.0   2.0  0.0  5.0  18.0  
58112  12.0   8.0   66  33.0   5.0   6.0  4.0  4.0  13.0  
59047   4.0   4.0   46  26.0   5.0   3.0  1.0  3.0   8.0  
60022   8.0   8.0   66  32.0   9.0   7.0  8.0  3.0  12.0  
61029   7.0   7.0   52  28.0   5.0   2.0  1.0  7.0  11.0  
61999   1.0   1.0   31  20.0   3.0   1.0  3.0  1.0  10.0  
62000   0.0   0.0    9   7.0   1.0   2.0  0.0  0.0   1.0  
62992   8.0   1.0   86  35.0   5.0   4.0  1.0  4.0  10.0  
64015   3.0   0.0   30  31.0   0.0   4.0  0.0  1.0   2.0  

[25 rows x 22 columns]

### Adding a `seasons` column

In [7]:
batting['seasons'] = batting['stint'].map(lambda x: x if x == 1 else 0)

### Adding a `year_rect` column

This eliminates multiple copies of the year for when players played with multiple teams in the same season. These two columns together will help when defining the `Era` column below.

In [8]:
def year_correct(row):
    """
    This function replaces multiple copies of the same year
    in a player's record with 0's.
    """
    if row['stint'] != 1:
        return 0
    else:
        return row['yearID']

In [9]:
batting['year_rect'] = batting.apply(year_correct, axis = 1)

In [10]:
batting[batting['playerID'] == 'rosepe01']

playerID  yearID  stint teamID lgID    G   AB    R    H  2B  ...    CS  \
43287  rosepe01    1963      1    CIN   NL  157  623  101  170  25  ...  15.0   
44033  rosepe01    1964      1    CIN   NL  136  516   64  139  13  ...  10.0   
44793  rosepe01    1965      1    CIN   NL  162  670  117  209  35  ...   3.0   
45567  rosepe01    1966      1    CIN   NL  156  654   97  205  38  ...   9.0   
46350  rosepe01    1967      1    CIN   NL  148  585   86  176  32  ...   6.0   
47085  rosepe01    1968      1    CIN   NL  149  626   94  210  42  ...   7.0   
47978  rosepe01    1969      1    CIN   NL  156  627  120  218  33  ...  10.0   
48920  rosepe01    1970      1    CIN   NL  159  649  120  205  37  ...   7.0   
49797  rosepe01    1971      1    CIN   NL  160  632   86  192  27  ...   9.0   
50693  rosepe01    1972      1    CIN   NL  154  645  107  198  31  ...   3.0   
51575  rosepe01    1973      1    CIN   NL  160  680  115  230  36  ...   7.0   
52484  rosepe01    1974      1    CIN   NL  163  652  110  185  45  ...   4.0   
53387  rosepe01    1975      1    CIN   NL  162  662  112  210  47  ...   1.0   
54275  rosepe01    1976      1    CIN   NL  162  665  130  215  42  ...   5.0   
55227  rosepe01    1977      1    CIN   NL  162  655   95  204  38  ...   4.0   
56194  rosepe01    1978      1    CIN   NL  159  655  103  198  51  ...   9.0   
57157  rosepe01    1979      1    PHI   NL  163  628   90  208  40  ...  11.0   
58112  rosepe01    1980      1    PHI   NL  162  655   95  185  42  ...   8.0   
59047  rosepe01    1981      1    PHI   NL  107  431   73  140  18  ...   4.0   
60022  rosepe01    1982      1    PHI   NL  162  634   80  172  25  ...   8.0   
61029  rosepe01    1983      1    PHI   NL  151  493   52  121  14  ...   7.0   
61999  rosepe01    1984      1    MON   NL   95  278   34   72   6  ...   1.0   
62000  rosepe01    1984      2    CIN   NL   26   96    9   35   9  ...   0.0   
62992  rosepe01    1985      1    CIN   NL  119  405   60  107  12  ...   1.0   
64015  rosepe01    1986      1    CIN   NL   72  237   15   52   8  ...   0.0   

        BB    SO   IBB   HBP   SH   SF  GIDP  seasons  year_rect  
43287   55  72.0   0.0   5.0  6.0  6.0   8.0        1       1963  
44033   36  51.0   0.0   2.0  3.0  1.0   6.0        1       1964  
44793   69  76.0   2.0   8.0  8.0  2.0  10.0        1       1965  
45567   37  61.0   3.0   1.0  7.0  1.0  12.0        1       1966  
46350   56  66.0   9.0   3.0  1.0  2.0   9.0        1       1967  
47085   56  76.0  15.0   4.0  2.0  4.0  11.0        1       1968  
47978   88  65.0  18.0   5.0  2.0  6.0  13.0        1       1969  
48920   73  64.0  10.0   2.0  0.0  4.0   7.0        1       1970  
49797   68  50.0  15.0   3.0  1.0  3.0   9.0        1       1971  
50693   73  46.0   4.0   7.0  2.0  2.0   7.0        1       1972  
51575   65  42.0   6.0   6.0  1.0  0.0  14.0        1       1973  
52484  106  54.0  14.0   5.0  1.0  6.0   9.0        1       1974  
53387   89  50.0   8.0  11.0  1.0  1.0  13.0        1       1975  
54275   86  54.0   7.0   6.0  0.0  2.0  17.0        1       1976  
55227   66  42.0   7.0   5.0  1.0  4.0   9.0        1       1977  
56194   62  30.0   6.0   3.0  2.0  7.0   8.0        1       1978  
57157   95  32.0  10.0   2.0  0.0  5.0  18.0        1       1979  
58112   66  33.0   5.0   6.0  4.0  4.0  13.0        1       1980  
59047   46  26.0   5.0   3.0  1.0  3.0   8.0        1       1981  
60022   66  32.0   9.0   7.0  8.0  3.0  12.0        1       1982  
61029   52  28.0   5.0   2.0  1.0  7.0  11.0        1       1983  
61999   31  20.0   3.0   1.0  3.0  1.0  10.0        1       1984  
62000    9   7.0   1.0   2.0  0.0  0.0   1.0        0          0  
62992   86  35.0   5.0   4.0  1.0  4.0  10.0        1       1985  
64015   30  31.0   0.0   4.0  0.0  1.0   2.0        1       1986  

[25 rows x 24 columns]

### Grouping By `playerID` and then Summing Will Show Career Totals

In [ ]:
players = batting.groupby('playerID').sum()

#### Hank Aaron, for example

In [ ]:
players[players.index == 'aaronha01']

### Adding a Batting Average Column

In [ ]:
players['avg'] = players['H'] / players['AB']
players[players.index == 'rosepe01']

### Adding a Slugging Average Column

In [ ]:
players['slg'] = (players['H'] + players['2B'] + 2 * players['3B'] + 3 * players['HR']) / players['AB']

In [ ]:
players.head()

## Finding the Worst Hitter

Beging by simply sorting according to batting average.

In [ ]:
players['avg'].sort_values()

But this is obviously not good enough. By this measure, any player who played in only a few games but never got a hit is at the top (or bottom) of the list. But let's think like a Bayesian!

### Bringing in Bayes

New plan: Find some average batting average to use as a *prior* probability distribution. Then treat the weight of a player's career as the *evidence* upon which to conditionalize. This will remove the players with very few career appearances from the top of our worst hitters list, since the prior (if we choose it appropriately) will dominate over the new evidence.  Inversely, players with long careers should have their actual numbers dominate over the prior.

Now getting a hit is a binary process (hit or no hit), so we can use the [beta-binomial distribution](https://en.wikipedia.org/wiki/Beta-binomial_distribution), since the beta distribution is a [conjugate prior](https://en.wikipedia.org/wiki/Conjugate_prior) for a binomial likelihood.

Taking advantage of conjugacy here means that I can simply add a player's career hits and at-bats to the prior's hits and at-bats to get an updated estimate of a player's batting average.

#### Setting the prior

I'm going to use .260 as an average MLB batting average.

In [ ]:
# Prior

successes = 26
failures = 74
alpha_prior = successes + 1
beta_prior = failures + 1

beta_dist = stats.beta(alpha_prior, beta_prior)

fig, ax = plt.subplots(figsize=(8, 5))
pvals = np.linspace(0, 1, 101)
prior = beta_dist.pdf(pvals)

ax.plot(pvals, prior, lw=3)
ax.set_xlabel('p', fontsize=16)
ax.set_ylabel('P(p | a,b)\n', fontsize=16)
ax.set_title(f'Beta PDF for alpha={alpha_prior}, beta={beta_prior}\n',
            fontsize=18)
plt.show()

#### Adding a Maximum A Posteriori Column

In [ ]:
# A:

players['MAP'] = (players['H'] + 26) / (players['AB'] + 100)

In [ ]:
players.sort_values('MAP', ascending=False).head()

In [ ]:
players = players.reset_index()

#### Adding an `Era` Column

In [ ]:
players = players.rename({'MAP': 'MAP_avg'}, axis=1)

players['Era'] = (players['year_rect'] / players['seasons']).astype(int)

Some players had perfect averages (going 1 for 1 in their careers or the like). Their new `MAP_avg` scores should be close to .260.

In [ ]:
players[players['avg'] == max(players['avg'])].head()

#### Experimenting with Different Numbers for the Prior's Hits and At-Bats

In [ ]:
def avg_prior(h, ab):
    """
    This function takes in a number of hits and a number of at-bats
    to use as prior values for the Bayesian MAP Method. It returns
    the top (worst) hitter according to the MAP average. The ratio
    of hits to at-bats should be (near) 26:100.
    """
    players['MAP_avg'] = (players['H'] + h) / (players['AB'] + ab)
    return players.sort_values('MAP_avg',
                               #ascending=False
                              ).head(10)['playerID']

The following code will find the worst hitter for prior values of at-bats between 10 and 5000, counting by tens.

In [ ]:
worst = []
for i in range(10, 5001, 10):
    worst.append(avg_prior(0.26 * i, i))

For just ten at-bats, the worst hitter is Ron Herbel.

In [ ]:
worst[0]

In [ ]:
worst[169]

In [ ]:
worst[170]

At precisely 1705 at-bats, the worst hitter switches from Bob Buhl to Bill Bergen.

In [ ]:
avg_prior(0.26 * 1704, 1704)

In [ ]:
avg_prior(0.26 * 1705, 1705)

In [ ]:
worst

In [ ]:
players[players['playerID'] == 'herbero01']

Ron Herbel: pitcher for the San Francisco Giants, San Diego Padres, New York Mets, and Atlanta Braves.

In [ ]:
players[players['playerID'] == 'chancde01']

Dean Chance: pitcher for the Los Angeles / California Angels, Minnesota Twins, Cleveland Indians, New York Mets, and Detroit Tigers.

In [ ]:
players[players['playerID'] == 'buhlbo01']

Bob Buhl: pitcher for the Milwaukee Braves, Chicago Cubs, and Philadelphia Phillies.

In [ ]:
players[players['playerID'] == 'bergebi01']

Bill Bergen: Cincinnati Reds and Brooklyn Superbas / Dodgers.

**Aside on the Best Hitters**

To find the *best* hitters, we can re-use our `avg_prior()` function, but simply include an `ascending=False` in the `.sort_values()` call.

In [ ]:
best = []
for i in range(10, 5001, 10):
    best.append(avg_prior(0.26 * i, i))

best

For almost all values of at-bats between 10 and 5000 (counting again by tens), the best hitter is Ty Cobb. But there are two names for 10 and 20 at-bats that I don't recognize.

In [ ]:
players[players['playerID'] == 'jansera01']

Ray Jansen: St. Louis Browns.

In [ ]:
players[players['playerID'] == 'sherlvi01']

Vince Sherlock: Brooklyn Dodgers.

In [ ]:
players[players['playerID'] == 'cobbty01']

### Looking for Non-Pitchers

Pitchers are specialized players who are generally given a bit of a break when it comes to hitting, and so really what I'm after is the worst hitter *who was not a pitcher*.

For these next cells I change the function to print out the top ten instead of just the top one.

In [ ]:
avg_prior(26, 100)

Ben Sheets: pitcher for the Milwaukee Brewers, Oakland Athletics, and Atlanta Braves.

Dick Ellsworth: pitcher for the Chicago Cubs, Philadelphia Phillies, Boston Red Sox, Cleveland Indians, and Milwaukee Brewers.

Bill Hands: pitcher for the San Francisco Giants, Chicago Cubs, Minnesota Twins, and Texas Rangers.

Al Leiter: pitcher for the New York Yankees, Toronto Blue Jays, Florida Marlins, and New York Mets.

Sandy Koufax: pitcher for the Brooklyn / Los Angeles Dodgers.

Brian Moehler: pitcher for the Detroit Tigers, Cincinnati Reds, Houston Astros, and Florida Marlins.

Roger Craig: pitcher for the Brooklyn / Los Angeles Dodgers, New York Mets, St. Louis Cardinals, Cincinnati Reds, and Philadelphia Phillies.

In [ ]:
avg_prior(52, 200)

Aaron Harang: pitcher for the Oakland Athletics, Cincinnati Reds, San Diego Padres, Los Angeles Dodgers, Seattle Mariners, New York Mets, Atlanta Braves, and Philadelphia Phillies.

John Burkett: pitcher for the San Francisco Giants, Florida Marlins, Texas Rangers, Atlanta Braves, and Boston Red Sox.

Nolan Ryan: pitcher for the New York Mets, California Angels, Houston Astros, and Texas Rangers.

In [ ]:
avg_prior(78, 300)

Mickey Lolich: pitcher for the Detroit Tigers, New York Mets, and San Diego Padres.

In [ ]:
avg_prior(104, 400)

Bob Friend: pitcher for the Pittsburgh Pirates, New York Yankees, and New York Mets.

Milt Pappas: pitcher for the Baltimore Orioles, Cincinnati Reds, Atlanta Braves, and Chicago Cubs.

In [ ]:
avg_prior(130, 500)

Jerry Koosman: pitcher for the New York Mets, Minnesota Twins, Chicago White Sox, and Philadelphia Phillies.

#### Bringing in pitching data

In [ ]:
pitching = pd.read_csv('../data/Pitching.csv')

#### Exploring the pitching data

In [ ]:
pitching['playerID'].values

In [ ]:
pitchers = pitching.groupby('playerID').sum()

Somewhat crudely, we'll look for the hitters who have no pitching record whatever.

In [ ]:
pure_bats = [player for player in players['playerID'] if player not in\
             pitching['playerID'].values]

In [ ]:
def also_pitch(x):
    """
    This function will be used to add a column to my
    players DataFrame that will indicate whether the
    player ever pitched.
    """
    if x['playerID'] in pure_bats:
        return 0
    else:
        return 1

In [ ]:
players['no_pitch'] = players.apply(also_pitch, axis = 1)

In [ ]:
pure_batters = players[players['no_pitch'] == 0].copy()

Now we'll just redo our previous MAP Method, but this time applying it only to the pure batters who didn't also pitch.

In [ ]:
def avg_prior_pure(h, ab):
    """
    This function mimics the MAP function from before.
    """
    pure_batters['MAP_avg'] = (pure_batters['H'] + h) / (pure_batters['AB'] + ab)
    return pure_batters.sort_values('MAP_avg',
                               # ascending = False
                              ).head(1)['playerID']

In [ ]:
worst = []
for i in range(10, 5001, 10):
    worst.append(avg_prior_pure(0.26 * i, i))

In [ ]:
worst

These names of pure hitters come up early in the list, but they correspond to very low numbers of at-bats as prior.

Skeeter Shelton: New York Yankees.

Ed Gastfield: Detroit Wolverines and Chicago White Stockings.

Mike Jordan: Pittsburgh Alleghenys.

John Humphries: New York Gothams and Washington Nationals.

It looks like **Bill Bergen** (who came up even before we eliminated the pitchers!) is our winner here. Let's find the exact number of at-bats where Bergen comes to the "top". Given the output from the preceding code, it must be somewhere between 130 and 140 at-bats.

In [ ]:
avg_prior_pure(0.26 * 132, 132)

In [ ]:
avg_prior_pure(0.26 * 133, 133)

***If we set our prior number of at-bats at 133 or more, Bill Bergen will count as the worst hitter in the history of the Major Leagues!***